src/kube-runtime/GOPATH/src/github.com/microsoft/runtime/cmd/exithandler/main.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/microsoft/runtime/pkg/aggregator"
+	"github.com/microsoft/runtime/pkg/logger"
+)
+
+var log *logger.Logger
+
+const defaultExitCode = 255
+
+func init() {
+	log = logger.NewLogger()
+}
+
+// This function will extract error summary to the specific file and print the exit code
+func main() {
+	argsWithoutProg := os.Args[1:]
+
+	if len(argsWithoutProg) < 5 {
+		log.Error("args is not valid")
+		os.Exit(defaultExitCode)
+	}
+
+	userExitCode, err := strconv.Atoi(argsWithoutProg[0])
+	if err != nil {
+		log.Error("user exit code is not an int value", err)
+		os.Exit(defaultExitCode)
+	}
+
+	userLog := argsWithoutProg[1]
+	runtimeErrorLog := argsWithoutProg[2]
+	aggFilePath := argsWithoutProg[3]
+	patternPath := argsWithoutProg[4]
+
+	logFiles := aggregator.LogFiles{}
+	logFiles.UserLog = userLog
+	logFiles.RuntimeErrorLog = runtimeErrorLog
+
+	exitInfo := &aggregator.RuntimeExitInfo{
+		Exitcode:           255,
+		OriginUserExitCode: userExitCode,
+	}
+
+	log.Info("start to generate the exit summary")
+	a, err := aggregator.NewErrorAggregator(&logFiles, log)
+	if err != nil {
+		log.Error("failed to create log aggregator", err)
+		goto DUMP_RESULT
+	}
+
+	err = a.LoadRuntimeErrorSpecs(patternPath)
+	if err != nil {
+		log.Error("load runtime error spec failed")
+		goto DUMP_RESULT
+	}
+	exitInfo, err = a.GenerateExitInfo(int(userExitCode))
+	if err != nil {
+		log.Error("failed to generate the exitInfo", err)
+	}
+
+DUMP_RESULT:
+	aggFile, err := os.Create(aggFilePath)
+	if err != nil {
+		log.Error("open aggregate file failed", err)
+		os.Exit(defaultExitCode)
+	}
+	defer aggFile.Close()
+
+	err = a.DumpExitSummary(exitInfo, aggFile)
+	if err != nil {
+		log.Error("dump exit info failed", err)
+		os.Exit(defaultExitCode)
+	}
+	log.Info("finish generating the exit summary")
+	fmt.Print(exitInfo.Exitcode)
+}
```

Comment:
When do you exit with exitInfo.Exitcode?

Reply:
Code exit at: https://github.com/microsoft/pai/blob/2061dd7d6431dae2970388babff2a352724eeea2/src/kube-runtime/src/runtime#L46

Reply:
It seems too tricky, can we unify the way to expose exitcode outside?

---
In reply to: [322047090](https://github.com/microsoft/pai/pull/3585#discussion_r322047090) [](ancestors = 322047090)

Reply:
I write like this for 2 reasons:
1. Currently, it run as a program. So, I want to use exit code to identify if the program run correctly.
2. I want to copy the aggerated info to log folder for debugging purpose. Since we set `set -o errexit`. Direct exit will terminate the main process

Ideally, this part should be executed as a function, then we can return the spec exit code and let main program exit with this code. But the main process is a shell script, it's not easy to do this unless we change the shell to a golang program.

A short term solution: exit with spec code. unset the `set -o errexit` in main shell. Then we can do some following actions such as log something, copy the aggregated exitInfo to log folder
A long term: use golang to rewrite the shell and let exitHandler be a function.

---
In reply to: [322048329](https://github.com/microsoft/pai/pull/3585#discussion_r322048329) [](ancestors = 322048329,322047090)